EPERM: operation not permitted, unlink '...node_modules\prisma\query_engine-windows.dll.node'

### Bug description

Following the [Getting Started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) guide, when it gets to the point of running the "migrate dev" command, the script tries to install some packages through npm (don't know which ones) and this causes the error in title.

### How to reproduce

1. Follow the [Getting started](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgres/) prisma guide up to and included the "npx prisma migrate dev --name init" command.
2. Get the following error:

```
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": PostgreSQL database "testprisma", schema "public" at "localhost:5432"

PostgreSQL database testprisma created at localhost:5432

The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20210909080511_init/
    └─ migration.sql

Your database is now in sync with your schema.

Running generate... (Use --skip-generate to skip the generators)
npm WARN hello-prisma@1.0.0 No description
npm WARN hello-prisma@1.0.0 No repository field.

npm ERR! code EPERM
npm ERR! syscall unlink
npm ERR! path C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node
npm ERR! errno -4048
npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'
npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!     errno: -4048,
npm ERR!     code: 'EPERM',
npm ERR!     syscall: 'unlink',
npm ERR!     path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node'
npm ERR!   },
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node',
npm ERR!   parent: 'hello-prisma'
npm ERR! }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\mbelletti\AppData\Roaming\npm-cache\_logs\2021-09-09T08_05_23_557Z-debug.log
Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.1
The batch file cannot be found.
```

The log file is the following:
[2021-09-09T08_05_23_557Z-debug.log](https://github.com/prisma/prisma/files/7134677/2021-09-09T08_05_23_557Z-debug.log)

After this, the node_modules\.bin directory no longer has the prisma executable, and the node_modules/prisma folder only contains the file that npm wasn't able to remove (query_engine-windows.dll.node).

### Expected behavior

The script finishes correctly.

### Prisma information

All the prisma configurations are the ones taken from the guide.

### Environment & setup

- OS: Windows 10
- Database: PostgreSQL 13
- Node.js version: 14.17.1 (using nvm)


### Prisma Version

```
Environment variables loaded from .env
prisma                  : 3.0.1
@prisma/client          : Not found
Current platform        : windows
Query Engine (Node-API) : libquery-engine 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : 2452cc6313d52b8b9a96999ac0e974d0aedf88db
Studio                  : 0.423.0
```

(I had to relaunch npm install to get prisma back)

I've tried right now by removing the node_modules directory, re-installing the packages with `npm install` and re-running `npx prisma migrate dev --name init`, and got the same result:

```
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": PostgreSQL database "testprisma", schema "public" at "localhost:5432"

Already in sync, no schema change or pending migration was found.

Running generate... (Use --skip-generate to skip the generators)
npm WARN hello-prisma@1.0.0 No description
npm WARN hello-prisma@1.0.0 No repository field.

npm ERR! code EPERM
npm ERR! syscall unlink
npm ERR! path C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node
npm ERR! errno -4048
npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'
npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'C:\Users\mbelletti\Documents\Projects\hello-prisma\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!     errno: -4048,
npm ERR!     code: 'EPERM',
npm ERR!     syscall: 'unlink',
npm ERR!     path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node'
npm ERR!   },
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path: 'C:\\Users\\mbelletti\\Documents\\Projects\\hello-prisma\\node_modules\\prisma\\query_engine-windows.dll.node',
npm ERR!   parent: 'hello-prisma'
npm ERR! }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\mbelletti\AppData\Roaming\npm-cache\_logs\2021-09-10T07_27_58_812Z-debug.log
Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.1
The batch file cannot be found.
```

I can confirm this problem:

```
C:\Users\Jan\Documents\throwaway\sethRepro>npx prisma migrate dev
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": SQLite database "fancy.db" at "file:./fancy.db"

SQLite database fancy.db created at file:./fancy.db

√ Enter a name for the new migration: ... init
The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20210910203934_init/
    └─ migration.sql

Your database is now in sync with your schema.

Running generate... (Use --skip-generate to skip the generators)
npm WARN sethRepro@1.0.0 No description
npm WARN sethRepro@1.0.0 No repository field.

npm ERR! code EPERM
npm ERR! syscall unlink
npm ERR! path C:\Users\Jan\Documents\throwaway\sethRepro\node_modules\prisma\query_engine-windows.dll.node
npm ERR! errno -4048
npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\Jan\Documents\throwaway\sethRepro\node_modules\prisma\query_engine-windows.dll.node'
npm ERR!  [OperationalError: EPERM: operation not permitted, unlink 'C:\Users\Jan\Documents\throwaway\sethRepro\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!   cause: [Error: EPERM: operation not permitted, unlink 'C:\Users\Jan\Documents\throwaway\sethRepro\node_modules\prisma\query_engine-windows.dll.node'] {
npm ERR!     errno: -4048,
npm ERR!     code: 'EPERM',
npm ERR!     syscall: 'unlink',
npm ERR!     path: 'C:\\Users\\Jan\\Documents\\throwaway\\sethRepro\\node_modules\\prisma\\query_engine-windows.dll.node'
npm ERR!   },
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path: 'C:\\Users\\Jan\\Documents\\throwaway\\sethRepro\\node_modules\\prisma\\query_engine-windows.dll.node',
npm ERR!   parent: 'sethRepro'
npm ERR! }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Jan\AppData\Roaming\npm-cache\_logs\2021-09-10T20_39_44_431Z-debug.log
Error: Command failed with exit code 4294963248: npm install -D prisma@3.0.2
The batch file cannot be found.
```

Here are my versions:

```
C:\Users\Jan\Documents\throwaway\sethRepro\prisma
λ node -v
v14.17.5

C:\Users\Jan\Documents\throwaway\sethRepro\prisma
λ npm -v
6.14.14
```

Interesting is what happened on `npx prisma -v`:
```
C:\Users\Jan\Documents\throwaway\sethRepro\prisma
λ npx prisma -v
Environment variables loaded from ..\.env
prisma                  : 3.0.2
@prisma/client          : Not found
Current platform        : windows
Query Engine (Node-API) : libquery-engine 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at ..\..\..\..\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at ..\..\..\..\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at ..\..\..\..\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt 2452cc6313d52b8b9a96999ac0e974d0aedf88db (at ..\..\..\..\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : 2452cc6313d52b8b9a96999ac0e974d0aedf88db
Studio                  : 0.423.0
Error: EPERM: operation not permitted, unlink 'C:\Users\Jan\AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\query_engine-windows.dll.node'
    at Object.unlinkSync (fs.js:1251:3)
    at fixWinEPERMSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:219:13)
    at rimrafSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:319:28)
    at C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:350:5
    at Array.forEach (<anonymous>)
    at rmkidsSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:349:26)
    at rmdirSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:342:7)
    at fixWinEPERMSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:217:5)
    at rimrafSync (C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:319:28)
    at C:\ProgramData\nvm\v14.17.5\node_modules\npm\node_modules\rimraf\rimraf.js:350:5
```

When running `npx prisma -v` it downloaded/installed something on the spot I think, so it might indeed be that the `prisma` installation somehow got corrupted. I have that sometimes but never could reproduce or understand it.

That my engine is located in `AppData\Roaming\npm-cache\_npx\29188\node_modules\prisma\node_modules\@prisma\engines\prisma-fmt-windows.exe` is also an indication that something is weird and this is not the actual `prisma` in my project, but a temporary one from `npx`.

Internal note:

This seems to be related to the Node-API library for Windows `query_engine-windows.dll.node`. When using the escape hatch to return to the binary engine, 
```
PRISMA_CLI_QUERY_ENGINE_TYPE=binary
PRISMA_CLIENT_ENGINE_TYPE=binary
```
(for example in your `.env` file), the error does _not_ occur like it does described above. Instead you get this one:

```
C:\Users\Jan\Documents\throwaway\9184>npx prisma generate
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma

> prisma@3.0.2 preinstall C:\Users\Jan\Documents\throwaway\9184\node_modules\prisma
> node scripts/preinstall-entry.js


> prisma@3.0.2 install C:\Users\Jan\Documents\throwaway\9184\node_modules\prisma
> node scripts/install-entry.js

npm notice save prisma is being moved from dependencies to devDependencies
npm WARN 9184@1.0.0 No description
npm WARN 9184@1.0.0 No repository field.

+ prisma@3.0.2
updated 1 package and audited 2 packages in 1.474s
found 0 vulnerabilities

npm WARN 9184@1.0.0 No description
npm WARN 9184@1.0.0 No repository field.

npm ERR! code ENOENT
npm ERR! syscall rename
npm ERR! path C:\Users\Jan\Documents\throwaway\9184\node_modules\@prisma\client
npm ERR! dest C:\Users\Jan\Documents\throwaway\9184\node_modules\@prisma\.client.DELETE
npm ERR! errno -4058
npm ERR! enoent ENOENT: no such file or directory, rename 'C:\Users\Jan\Documents\throwaway\9184\node_modules\@prisma\client' -> 'C:\Users\Jan\Documents\throwaway\9184\node_modules\@prisma\.client.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Jan\AppData\Roaming\npm-cache\_logs\2021-09-20T19_03_16_214Z-debug.log
Error: Command failed with exit code 4294963238: npm install @prisma/client@3.0.2
```

--- a/packages/sdk/src/predefinedGeneratorResolvers.ts
+++ b/packages/sdk/src/predefinedGeneratorResolvers.ts
@@ -97,7 +97,7 @@ export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
         console.info(`✔ Created ${chalk.bold.green('./package.json')}`)
       }
 
-      await installPackage(baseDir, `-D prisma@${version ?? 'latest'}`)
+      // await installPackage(baseDir, `-D prisma@${version ?? 'latest'}`)
       await installPackage(baseDir, `@prisma/client@${version ?? 'latest'}`)
 
       // resolvePkg has caching, so we trick it not to do it 👇

--- a/packages/sdk/src/predefinedGeneratorResolvers.ts
+++ b/packages/sdk/src/predefinedGeneratorResolvers.ts
@@ -51,6 +51,42 @@ async function findPrismaClientDir(baseDir: string) {
   return resolvedClientDir
 }
 
+function getPackageJsonPath(baseDir: string): string | undefined {
+  // TODO: these locations (moved from an if statement below in 'prisma-client-js' resolver)
+  // don't look right in general case, we need to find the nearest package.json relative to baseDir.
+  const currentDirectory = path.join(process.cwd(), 'package.json')
+  const parentDirectory = path.join(process.cwd(), '..', 'package.json')
+
+  for (const dir of [currentDirectory, parentDirectory]) {
+    if (fs.existsSync(dir)) {
+      return dir
+    }
+  }
+
+  return undefined
+}
+
+async function isPrismaCliInstalledInCurrentPackage(baseDir: string): Promise<boolean> {
+  const packageJsonPath = getPackageJsonPath(baseDir)
+
+  if (packageJsonPath === undefined) {
+    return false
+  }
+
+  const cliDir = await resolvePkg('prisma', {
+    basedir: baseDir,
+    preserveSymlinks: true,
+  })
+
+  if (cliDir === undefined) {
+    return false
+  }
+
+  const relativeDir = path.relative(path.dirname(packageJsonPath), cliDir)
+
+  return !relativeDir.startsWith('..')
+}
+
 export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
   photonjs: () => {
     throw new Error(`Oops! Photon has been renamed to Prisma Client. Please make the following adjustments:
@@ -75,10 +111,7 @@ export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
     await checkTypeScriptVersion()
 
     if (!prismaClientDir && !process.env.PRISMA_GENERATE_SKIP_AUTOINSTALL) {
-      if (
-        !fs.existsSync(path.join(process.cwd(), 'package.json')) &&
-        !fs.existsSync(path.join(process.cwd(), '../package.json'))
-      ) {
+      if (!getPackageJsonPath(baseDir)) {
         // Create default package.json
         const defaultPackageJson = `{
   "name": "my-prisma-project",
@@ -97,7 +130,10 @@ export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
         console.info(`✔ Created ${chalk.bold.green('./package.json')}`)
       }
 
-      // await installPackage(baseDir, `-D prisma@${version ?? 'latest'}`)
+      if ((await isPrismaCliInstalledInCurrentPackage(baseDir)) === false) {
+        await installPackage(baseDir, `-D prisma@${version ?? 'latest'}`)
+      }
+
       await installPackage(baseDir, `@prisma/client@${version ?? 'latest'}`)
 
       // resolvePkg has caching, so we trick it not to do it 👇

--- a/packages/sdk/src/predefinedGeneratorResolvers.ts
+++ b/packages/sdk/src/predefinedGeneratorResolvers.ts
@@ -51,42 +51,6 @@ async function findPrismaClientDir(baseDir: string) {
   return resolvedClientDir
 }
 
-function getPackageJsonPath(baseDir: string): string | undefined {
-  // TODO: these locations (moved from an if statement below in 'prisma-client-js' resolver)
-  // don't look right in general case, we need to find the nearest package.json relative to baseDir.
-  const currentDirectory = path.join(process.cwd(), 'package.json')
-  const parentDirectory = path.join(process.cwd(), '..', 'package.json')
-
-  for (const dir of [currentDirectory, parentDirectory]) {
-    if (fs.existsSync(dir)) {
-      return dir
-    }
-  }
-
-  return undefined
-}
-
-async function isPrismaCliInstalledInCurrentPackage(baseDir: string): Promise<boolean> {
-  const packageJsonPath = getPackageJsonPath(baseDir)
-
-  if (packageJsonPath === undefined) {
-    return false
-  }
-
-  const cliDir = await resolvePkg('prisma', {
-    basedir: baseDir,
-    preserveSymlinks: true,
-  })
-
-  if (cliDir === undefined) {
-    return false
-  }
-
-  const relativeDir = path.relative(path.dirname(packageJsonPath), cliDir)
-
-  return !relativeDir.startsWith('..')
-}
-
 export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
   photonjs: () => {
     throw new Error(`Oops! Photon has been renamed to Prisma Client. Please make the following adjustments:
@@ -111,7 +75,12 @@ export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
     await checkTypeScriptVersion()
 
     if (!prismaClientDir && !process.env.PRISMA_GENERATE_SKIP_AUTOINSTALL) {
-      if (!getPackageJsonPath(baseDir)) {
+      // TODO: `prisma generate` may be called deeper than one subdirectory from
+      // the package root.
+      if (
+        !fs.existsSync(path.join(process.cwd(), 'package.json')) &&
+        !fs.existsSync(path.join(process.cwd(), '../package.json'))
+      ) {
         // Create default package.json
         const defaultPackageJson = `{
   "name": "my-prisma-project",
@@ -130,8 +99,10 @@ export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
         console.info(`✔ Created ${chalk.bold.green('./package.json')}`)
       }
 
-      if ((await isPrismaCliInstalledInCurrentPackage(baseDir)) === false) {
-        await installPackage(baseDir, `-D prisma@${version ?? 'latest'}`)
+      const prismaCliDir = await resolvePkg('prisma', { basedir: baseDir })
+
+      if (!prismaCliDir) {
+        await installPackage(baseDir, `prisma@${version ?? 'latest'}`, 'dev')
       }
 
       await installPackage(baseDir, `@prisma/client@${version ?? 'latest'}`)
@@ -142,9 +113,9 @@ export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
       if (!prismaClientDir) {
         throw new Error(
           `Could not resolve @prisma/client despite the installation that we just tried.
-Please try to install it by hand with ${chalk.bold.greenBright('npm install @prisma/client')} and rerun ${chalk.bold(
-            getCommandWithExecutor('prisma generate'),
-          )} 🙏.`,
+Please try to install it by hand with ${chalk.bold.greenBright(
+            `${getAddPackageCommandName(baseDir)} @prisma/client`,
+          )} and rerun ${chalk.bold(getCommandWithExecutor('prisma generate'))} 🙏.`,
         )
       }
 
@@ -172,10 +143,23 @@ Please try to install it with ${chalk.bold.greenBright('npm install @prisma/clie
   },
 }
 
-async function installPackage(baseDir: string, pkg: string): Promise<void> {
-  const yarnUsed = hasYarn(baseDir) || hasYarn(path.join(baseDir, '..'))
+function isYarnUsed(baseDir: string): boolean {
+  // TODO: this may give false results for Yarn workspaces, implement proper detection.
+  return hasYarn(baseDir) || hasYarn(path.join(baseDir, '..'))
+}
 
-  const cmdName = yarnUsed ? 'yarn add' : 'npm install'
+function getAddPackageCommandName(baseDir: string, dependencyType?: 'dev'): string {
+  let command = isYarnUsed(baseDir) ? 'yarn add' : 'npm install'
+
+  if (dependencyType === 'dev') {
+    command += ' -D'
+  }
+
+  return command
+}
+
+async function installPackage(baseDir: string, pkg: string, dependencyType?: 'dev'): Promise<void> {
+  const cmdName = getAddPackageCommandName(baseDir, dependencyType)
 
   await execa.command(`${cmdName} ${pkg}`, {
     cwd: baseDir,

--- a/packages/sdk/src/predefinedGeneratorResolvers.ts
+++ b/packages/sdk/src/predefinedGeneratorResolvers.ts
@@ -101,6 +101,24 @@ export const predefinedGeneratorResolvers: PredefinedGeneratorResolvers = {
 
       const prismaCliDir = await resolvePkg('prisma', { basedir: baseDir })
 
+      // Automatically installing the packages with Yarn on Windows won't work because
+      // Yarn will try to unlink the Query Engine DLL, which is currently being used.
+      // See https://github.com/prisma/prisma/issues/9184
+      if (process.platform === 'win32' && isYarnUsed(baseDir)) {
+        const hasCli = (s: string) => (prismaCliDir !== undefined ? s : '')
+        const missingCli = (s: string) => (prismaCliDir === undefined ? s : '')
+
+        throw new Error(
+          `Could not resolve ${missingCli(`${chalk.bold('prisma')} and `)}${chalk.bold(
+            '@prisma/client',
+          )} in the current project. Please install ${hasCli('it')}${missingCli('them')} with ${missingCli(
+            `${chalk.bold.greenBright(`${getAddPackageCommandName(baseDir, 'dev')} prisma`)} and `,
+          )}${chalk.bold.greenBright(`${getAddPackageCommandName(baseDir)} @prisma/client`)}, and rerun ${chalk.bold(
+            getCommandWithExecutor('prisma generate'),
+          )} 🙏.`,
+        )
+      }
+
       if (!prismaCliDir) {
         await installPackage(baseDir, `prisma@${version ?? 'latest'}`, 'dev')
       }

--- a/packages/sdk/src/predefinedGeneratorResolvers.ts
+++ b/packages/sdk/src/predefinedGeneratorResolvers.ts
@@ -144,7 +144,9 @@ Please try to install it with ${chalk.bold.greenBright('npm install @prisma/clie
 }
 
 function isYarnUsed(baseDir: string): boolean {
-  // TODO: this may give false results for Yarn workspaces, implement proper detection.
+  // TODO: this may give false results for Yarn workspaces or when the schema is
+  // in a non-standard location, implement proper detection.
+  // Possibly related: https://github.com/prisma/prisma/discussions/10488
   return hasYarn(baseDir) || hasYarn(path.join(baseDir, '..'))
 }
 

